tests/package.json

```diff
@@ -27,6 +27,9 @@
     "mini:default": "cd .. && npx esbuild ./src/*js --minify --allow-overwrite --outdir=src",
     "mini:win32": "cd .. && cd src && forfiles /m \"*.js\" /c \"cmd /c npx esbuild @file --minify --allow-overwrite --outfile=@file\"",
     "mini": "run-script-os",
+    "embed:cp-refd": "node --input-type=module -e \"import fs from 'fs';fs.copyFileSync('../node_modules/resolvewithplus/resolvewithplus.js', '../src/resolvewithplus.js')\"",
+    "embed:update-refs": "node --input-type=module -e \"import fs from 'fs';let p='../src/esmockModule.js';fs.writeFileSync(p, fs.readFileSync(p, 'utf-8').replace(/'resolvewithplus'/, '\\'./resolvewithplus.js\\''))\"",
+    "embed": "npm run embed:update-refs && npm run embed:cp-refd",
```

Comment:
not sure I've gotten your point: `npm run embed` should be run before publish, but I didn't see it was added.

I had a concern: the code in here is really fallible, and hard to debug. can we just copy the `resolvewithplus` to a folder, and provide a script to update it to the newer version when you want?

I made a similar thing in `eslint-plugin-n`:
* https://github.com/eslint-community/eslint-plugin-n/blob/master/scripts/convert-pure-esm-to-cjs.js
* https://github.com/eslint-community/eslint-plugin-n/blob/master/lib/converted-esm/import-meta-resolve.js

Reply:
`npm run embed` is used by `npm run test:all-ci`, also used by the prepublish script,

combining parts of the main and test package.json scripts,
```json
{
    "embed:cp-refd": "node --input-type=module -e \"import fs from 'fs';fs.copyFileSync('../node_modules/resolvewithplus/resolvewithplus.js', '../src/resolvewithplus.js')\"",
    "embed:update-refs": "node --input-type=module -e \"import fs from 'fs';let p='../src/esmockModule.js';fs.writeFileSync(p, fs.readFileSync(p, 'utf-8').replace(/'resolvewithplus'/, '\\'./resolvewithplus.js\\''))\"",
    "embed": "npm run embed:update-refs && npm run embed:cp-refd",
    "mini:pkg": "npm pkg delete scripts devDependencies",
    "test:all-ci": "npm run mini && npm run embed && npm run test:all",
    "test-ci": "npm run test:install && npm run test:all-ci",
    "prepublishOnly": "npm run lint && npm run test-ci && npm run mini:pkg"
}
```
If storing resolvewithplus directly inside the src folder is preferred that is OK with me. The above are complicated and ugly, but they are verified at the unit-test pipeline.


[0]: https://github.com/iambumblehead/esmock/blob/master/package.json#L79

Reply:
a little ugly but not too terrible (for me)

_embed:cp-refd_
```javascript
import fs from 'fs';
fs.copyFileSync('../node_modules/resolvewithplus/resolvewithplus.js', '../src/resolvewithplus.js')
```
_embed:update-refs_
```javascript
import fs from 'fs';
let p='../src/esmockModule.js';
fs.writeFileSync(p, fs.readFileSync(p, 'utf-8').replace(/'resolvewithplus'/, '\'./resolvewithplus.js\''))

Reply:
I slightly prefer storing in the src, to avoid the magic like `embed:cp-refd`/`embed:update-refs`, but let's see what others think. :)

Reply:
if one trusts these lines, they allow one to develop esmock without needing to manually sync resolvewithplus and without needing to add extra files. esmock can be developed and used with regular npm, while unit-tests and prepublishOnly automaticaly get the embedded result.

Reply:
my internet connection is slow I did not see your reply when adding my last message

Reply:
@koshic @tripodsan what are your opinions? :)

Reply:
Using current approach, it remains easy to reference [new resolvewithplus branches.][0] Simply update package.json and use npm as usual. No need to sync a file or include sync file changes in the PR,
```json
{
  "dependencies": {
    "resolvewithplus": "iambumblehead/resolvewithplus#return-encoded-url-same-as-import.meta.resolve"
  }
}
```
The other approach, committing resolvewithplus.js into the repo, creates work and adds noise to the PR, where sync file changes become an added part of the PR.

[0]: https://github.com/iambumblehead/resolvewithplus/pull/40

Reply:
The main question - what is our goal in terms of package internal structure / development workflow / etc. What do I mean? 

Build / compile / bundle strep is not something new, and my personal preference (due to a multiple factors) is to use resolvewithplus package as devDependency and just call 'build' before publishing. 'Build' can contains absolutely anything, from 'cp' to 'webpack'. But esbuild is better ). This pattern is transparent and not introduce any kind of magic.

BUT if you want to avoid external tools usage, specially designed tests or even have compilerphobia, it may make sense. And 'cp' pattern is not so ugly for me. You know, @aladdin-add knows, I know - that's enough.

PS Sure, in my current company developer will be forced to rewrite 'that crap' (especially if somebody adds me to the review for old times sake), just because he is not alone and nobody wants to understand his magic in the future.

Reply:
> Build / compile / bundle strep is not something new, and my personal preference (due to a multiple factors) is to use resolvewithplus package as devDependency and just call 'build' before publishing.

What do you think of this idea @aladdin-add @koshic?
  1. move resolvewithplus to devDependency, per @koshic
  2. update esbuild commands to output single bundle file, per @koshic 
  3. esbuild output saved to src/ with the package-exported name 'esmockLoader.js'. This way, esmock does not need a 'dist' or 'build' folder and can continue to be cloned from git and used as-is without build step,
  4. other files are deleted

final output would look like this,
```console
src/
├── esmock.d.ts
└── esmockLoader.js
```


[0]: https://github.com/iambumblehead/esmock/blob/master/tests/package.json#L27

Reply:
esbuild is rigid and may not allow us to output a file the same name, which brings us back to the problem of missing rm and cp commands in windows-latest pipelines

Reply:
Let me play with master & esbuild, it will take ~30m

Reply:
Place the following code in build.js into the package root:

```js
import { fileURLToPath } from "node:url";
import { build } from "esbuild";

await build({
  outdir: ".",
  entryPoints: ["esmockModule.js"],
  bundle: true,
  // can be globbed with any glob package like ['src/*', '!/src/esmockModule.js']
  external: ["./esmockErr.js", "./esmockCache.js"],
  // can be ommited via 'cd src' before
  absWorkingDir: fileURLToPath(new URL("src", import.meta.url)),
  format: "esm",
  platform: "node",
  // to be sure that nothing was changed, very safe to remove
  target: "node18",
  allowOverwrite: true,
});

```

It can be simplified via running it from 'src' dir, or called via command line with the same options.

Reply:
I'll be un-available to respond until later this evening. Thanks for the discussion and script demo

Reply:
I'm confused a bit )

My proposal (and code) is about adding one step right before package publishing - to bundle 'resolvewith' into esmockModule.js. There is no reason to run bundle step for tests, 'resolvewith' should stay as external dependency and they should work with original esmockModule.js.

Reply:
The final, bundled, minified version is unit-tested to ensure published sources pass all tests, before they are published. If tests fail, the publish step is aborted.

Reply:
Thx, now I catch it   we misunderstood each other )

I proposed 'precision strike' - modify only one file (esmockModule.js) during 'bundle' step to keep your current flow. That step can be executed before tests (and followed publish), as you prefer.

But you decided to solve this task more radically - to bundle all .js files into one, then test & publish resulting bundle (esmockLoader.js) plus types. Right?

Reply:
yes :) https://github.com/iambumblehead/esmock/pull/181#discussion_r1000842074
> update esbuild commands to output single bundle file, per @koshic

final output would look like this,
```console
src/
├── esmock.d.ts
└── esmockLoader.js
```

Reply:
Nice ) Only one note: you don't need 'src' anymore in the final structure. Types can be moved the at root without any concerns, bundle - with minor tweaks. What do you think? As I see, you really like minimal style )

Reply:
If we generate output at src, we can continue pointing [exports.import][0] at the src location. The benefit: the project can be cloned and used right away without build step.

It would be good to export esmock rather than esmockLoader maybe in a follow up PR.


[0]: https://github.com/iambumblehead/esmock/blob/master/package.json#L10-L13

Reply:
I'll update the PR again in a day or two

Reply:
> If we generate output at src, we can continue pointing [exports.import](https://github.com/iambumblehead/esmock/blob/master/package.json#L10-L13) at the src location. The benefit: the project can be cloned and used right away without build step.

Note for the next PR: publishConfig can keep this benefit and provide different exports for different situations.

Reply:
@koshic @aladdin-add unusual "ref" scripts are removed and esbuild "bundle" is used. What do you think?
```console
npx esbuild esmockLoader.js \
  --minify \
  --bundle \
  --allow-overwrite \
  --platform=node \
  --format=esm \
  --outfile=esmockLoader.js
```

Reply:
thank you for sparing me your time so far

Reply:
I'd prefer two entry different points in package.json - one for '--loader', one for 'import esmock' inside test code. Why?

To avoid mixing code executed in different sandboxes for different purposes, and possible side-effects.

So, ie may look like '--loader esmock/loader' and 'import {xxx} from "esmock"' from regular code:
```
exports: {
  "types": "...",
  "import": {
    ".": "./dist/main.js",
    "./loader": "./dist/loader.js",
  }
}
```

There is one thing - the more the build result differs from the source files, the less sense it makes to test this result. As an example, for C++ code you will never write tests to check .exe file, right?

Here is the same: bundle(s) is not the best source for unit tests (that's why I use 'dist' dir in my example). It can be tested in kind of e2e tests ('loaded without errors', 'has exports A & B', etc.), but regular unit tests is about raw source by many reasons: you can have multiple output formats, project can migrate to TS, you may want to change bundler (and output) without touching unit tests, and so on.

So, I propose to have independent 'source with unit tests' & 'bundled code with e2e tests' layers. And it a good point for discussion - but in a different PR. This is PR is ready to merge.

Reply:
> I'd prefer two different entry points in package.json - one for '--loader', one for 'import esmock' inside test code. Why?

The current interface using `--loader=esmock` and`import esmock from ''esmock"` is easier for users and documentation. 


> To avoid mixing code executed in different sandboxes for different purposes, and possible side-effects.

The loader will [run in a separate thread][0] soon and all tests currently pass for supported versions of node.


> There is one thing - the more the build result differs from the source files, the less sense it makes to test this result.

The un-processed sources are tested at the [code coverage pipeline,][1] but only on the node18 image. This PR (and other PRs) did and do encounter bugs related to bundling and minification and it is useful to test these files before publishing them.


> As an example, for C++ code you will never write tests to check .exe file, right?

This is true generally, but there is a unique context for esmock. Functional tests give coverage that is nearly 100% and exporting and unit-testing each function is not needed imho. Input and output results are predictable.



[0]: https://github.com/nodejs/node/pull/44710
[1]: https://github.com/iambumblehead/esmock/actions/runs/3303674831/jobs/5451893808#step:7:558

Reply:
@aladdin-add this PR waits or you approval :)